runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaPipelineResult.java

```diff
@@ -52,40 +50,41 @@ public State getState() {
   }
 
   @Override
-  public State cancel() throws IOException {
+  public State cancel() {
     runner.kill(app);
-
-    //TODO: runner.waitForFinish() after SAMZA-1653 done
-    return getState();
+    return waitUntilFinish();
   }
 
   @Override
   public State waitUntilFinish(Duration duration) {
-    //TODO: SAMZA-1653
-    throw new UnsupportedOperationException(
-        "waitUntilFinish(duration) is not supported by the SamzaRunner");
+    try {
+      runner.waitForFinish(java.time.Duration.ofMillis(duration.getMillis()));
+    } catch (Exception e) {
+      throw new Pipeline.PipelineExecutionException(e);
+    }
+
+    final StateInfo stateInfo = getStateInfo();
+    if (stateInfo.state == State.FAILED) {
+      throw stateInfo.error;
+    }
+
+    return stateInfo.state;
   }
 
   @Override
   public State waitUntilFinish() {
-    if (runner instanceof LocalApplicationRunner) {
-      try {
-        ((LocalApplicationRunner) runner).waitForFinish();
-      } catch (Exception e) {
-        throw new Pipeline.PipelineExecutionException(e);
-      }
-
-      final StateInfo stateInfo = getStateInfo();
-      if (stateInfo.state == State.FAILED) {
-        throw stateInfo.error;
-      }
-
-      return stateInfo.state;
-    } else {
-      // TODO: SAMZA-1653 support waitForFinish in remote runner too
-      throw new UnsupportedOperationException(
-          "waitUntilFinish is not supported by the SamzaRunner when running remotely");
+    try {
```

Comment:
I don't understand the duplication here. Why can't we combine them?

Reply:
Sure. let me clean up the code to share the logic.